database/src/service/read.rs

```diff
@@ -228,57 +317,191 @@ fn map_request(
+            height: u64::from(output.height),
+            time_lock,
+            key,
+            commitment,
+        };
+
+        Ok((amount_index, output_on_chain))
+    };
+
+    let map = map
+        .into_par_iter()
+        .map(|(amount, amount_index_set)| {
+            Ok((
+                amount,
+                amount_index_set
+                    .into_par_iter()
+                    .map(|amount_index| inner_map(amount, amount_index))
+                    .collect::<Result<HashMap<AmountIndex, OutputOnChain>, RuntimeError>>()?,
+            ))
+        })
+        .collect::<Result<HashMap<Amount, HashMap<AmountIndex, OutputOnChain>>, RuntimeError>>()?;
+
+    Ok(Response::Outputs(map))
 }
 
 /// [`ReadRequest::NumberOutputsWithAmount`].
-/// TODO
 #[inline]
-#[allow(clippy::needless_pass_by_value)] // TODO: remove me
-fn number_outputs_with_amount(env: &Arc<ConcreteEnv>, vec: Vec<Amount>) -> ResponseResult {
-    todo!()
+fn number_outputs_with_amount(env: &ConcreteEnv, amounts: Vec<Amount>) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let (tx_ro, tables) = set_tx_ro_and_tables!(env, env_inner);
+
+    let map = amounts
+        .into_par_iter()
+        .map(|amount| {
+            let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+
+            match tables.num_outputs().get(&amount) {
+                // INVARIANT: #[cfg] @ lib.rs asserts `usize == u64`
+                #[allow(clippy::cast_possible_truncation)]
+                Ok(count) => Ok((amount, count as usize)),
+                Err(e) => Err(e),
+            }
+        })
+        .collect::<Result<HashMap<Amount, usize>, RuntimeError>>()?;
+
+    Ok(Response::NumberOutputsWithAmount(map))
 }
 
 /// [`ReadRequest::CheckKIsNotSpent`].
 #[inline]
-#[allow(clippy::needless_pass_by_value)] // TODO: remove me
-fn check_k_is_not_spent(env: &Arc<ConcreteEnv>, set: HashSet<KeyImage>) -> ResponseResult {
-    todo!()
+fn check_k_is_not_spent(env: &ConcreteEnv, key_images: HashSet<KeyImage>) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let (tx_ro, tables) = set_tx_ro_and_tables!(env, env_inner);
+
+    let key_image_exists = |key_image| {
+        let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+        key_image_exists(&key_image, tables.key_images())
+    };
+
+    match key_images
+        .into_par_iter()
+        .map(key_image_exists)
+        // If the result is either:
+        // `Ok(true)` => a key image was found, return early
+        // `Err` => an error was found, return early
+        //
+        // Else, `Ok(false)` will continue the iterator.
+        .find_any(|result| !matches!(result, Ok(false)))
+    {
+        None | Some(Ok(false)) => Ok(Response::CheckKIsNotSpent(true)), // Key image was NOT found.
+        Some(Ok(true)) => Ok(Response::CheckKIsNotSpent(false)),        // Key image was found.
+        Some(Err(e)) => Err(e), // A database error occurred.
+    }
```

Comment:
`Response::CheckKIsNotSpent(true)` means the key image doesn't exist, right?

Reply:
Think adding an `enum cuprate_types::Exist { Does, DoesNot }` or similar would be nice btw.

Reply:
IIRC I switched it to this in the document consensus PR so for now just put it as a big TODO to make sure, if this is incorrect all txs will fail verification anyway.